File: scripts/verify_eight_gates.ts
import assert from 'node:assert/strict'
import { evaluateEightGates } from '../netlify/functions/lib/eight-gates'

const profile = (overrides: any = {}) => ({ canonical_source: 'fixture', adapter_version: 'fixture:v1', semantic_version: 'source-contract:v2', certified: true, automation_enabled: true, auto_enabled: false, distribution_policy: { web_catalog_allowed: false, google_jobs_distribution_allowed: false, third_party_job_distribution_allowed: false, source_attribution_required: false }, ...overrides })
const row = (overrides: any = {}) => ({ description: 'Useful description '.repeat(8), organization: 'Org', location: 'Asunción', country_code: 'PY', application_url: 'https://apply.test/1', source_url: 'https://source.test/1', remote_scope: 'ONSITE', ...overrides })
const run = (overrides: any = {}) => ({ status: 'healthy', found_count: 1, finished_at: '2026-09-15T00:00:00Z', extraction_metrics: {}, ...overrides })

const healthy = evaluateEightGates(profile(), [row()], run(), [], [])
assert.equal(healthy.gates[3].status, 'PASS')
assert.equal(healthy.gates[6].status, 'NOT_APPLICABLE')
assert.equal(healthy.gates[6].reason_code, 'AUTO_DISABLED_BY_POLICY')
assert.equal(healthy.gates[7].reason_code, 'RESTRICTED_BY_POLICY')

const talent = evaluateEightGates(profile({ canonical_source: 'talentcom' }), [row({ description: '', source_url: '' })], run(), [], [])
assert.equal(talent.gates[3].reason_code, 'CONTENT_INSUFFICIENT')
assert.equal(talent.gates[5].reason_code, 'RUNTIME_PASS_DATA_HEALTH_FAIL')
assert.equal(talent.gates[1].reason_code, 'DETAIL_UNPROVEN') // UNKNOWN remains unknown, never NOT_PUBLISHED

const lost = evaluateEightGates(profile({ auto_enabled: true, distribution_policy: { web_catalog_allowed: true } }), [row({ description: '' })], run({ extraction_metrics: { eight_gates: { gate_2: { metrics: { fields: { description: { extracted: 1 } } } } } } }), [], [])
assert.equal(lost.gates[4].reason_code, 'DESCRIPTION_LOST_BEFORE_PERSISTENCE')
assert.equal(lost.gates[6].reason_code, 'BLOCKED_BY_PREVIOUS_GATE')
assert.equal(lost.gates[6].status, 'NOT_EVALUATED')
assert.equal(lost.gates[7].reason_code, 'BLOCKED_BY_PREVIOUS_GATE')
assert.equal(lost.gates[7].evidence.blocked_by_gate, 4)

const ready = evaluateEightGates(profile({ canonical_source: 'unjobs', auto_enabled: true, distribution_policy: { web_catalog_allowed: true, google_jobs_distribution_allowed: true, third_party_job_distribution_allowed: true, source_attribution_required: true } }), [row()], run(), [], [])
assert.equal(ready.gates[6].status, 'PASS')
assert.equal(ready.gates[6].reason_code, 'READY_FOR_AUTOMATION')

const parser = evaluateEightGates(profile(), [row()], run({ extraction_metrics: { eight_gates: { gate_2: { status: 'FAIL', reason_code: 'DETAIL_UNAVAILABLE', metrics: {} } } } }), [], [])
assert.equal(parser.gates[1].status, 'FAIL')

const himalayas = evaluateEightGates(profile({ canonical_source: 'himalayas', distribution_policy: { web_catalog_allowed: true, google_jobs_distribution_allowed: false, third_party_job_distribution_allowed: false, source_attribution_required: true } }), [row()], run(), [], [])
assert.equal(himalayas.gates[6].reason_code, 'AUTO_DISABLED_BY_POLICY')
assert.equal(himalayas.gates[7].status, 'WARNING')
assert.equal(himalayas.gates[7].reason_code, 'PARTIALLY_ALLOWED')
assert.equal(himalayas.gates[7].metrics.surfaces.organic_seo.state, 'POLICY_NOT_DEFINED')

const uncertified = evaluateEightGates(profile({ certified: false }), [row()], run(), [], [])
assert.equal(uncertified.gates[6].status, 'NOT_APPLICABLE')
assert.equal(uncertified.gates[6].reason_code, 'SOURCE_NOT_CERTIFIED')
assert.notEqual(uncertified.gates[6].status, 'FAIL')

const policyUndefined = evaluateEightGates(profile({ distribution_policy: undefined }), [row()], run(), [], [])
assert.equal(policyUndefined.gates[7].status, 'NOT_EVALUATED')
assert.equal(policyUndefined.gates[7].reason_code, 'POLICY_NOT_DEFINED')
console.log('PASS eight gates UI projection: previous-gate block, AUTO policy, partial distribution, undefined SEO/AEO/GEO, uncertified source, non-failing restrictions')

File: scripts/verify_eight_gates.py
"""Fixtures for Source Intelligence — Eight Gates run evidence."""
from pathlib import Path
import sys

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT / "scrapers"))
from source_adapters import AdapterResult
from source_evidence import eight_gates_run_evidence, runtime_telemetry


def detail(**values):
    base = dict(source="fixture", adapter_version="fixture:v1", source_url="https://source.test/1", apply_url="https://apply.test/1", title="Role", organization="Org", description="Useful role description " * 8, location="Asunción", country_code="PY", remote_scope="ONSITE", deadline="2026-12-31", employment_type="job", source_status=200)
    base.update(values)
    return AdapterResult(**base)


runtime = runtime_telemetry(provider_health="HEALTHY", coverage_complete=True, coverage_stop_reason=None, found=1, valid=1)
assert runtime["operational_health"]["status"] == "HEALTHY"


healthy = eight_gates_run_evidence(source="himalayas", adapter_version="fixture:v1", metrics={"found": 1, "detail_pages_success": 1, "parsed": 1, "coverage_complete": True}, details=[detail()], summary={"valid": 1, "rejected": 0})
assert healthy["gate_1"]["status"] == "PASS" and healthy["gate_2"]["status"] == "PASS"
assert healthy["gate_2"]["metrics"]["fields"]["description"]["state"] == "EXTRACTED"

# Runtime can be reachable while data is unusable: the detail gate warns rather
# than pretending a blank expected field was published by the provider.
data_bad = eight_gates_run_evidence(source="talentcom", adapter_version="fixture:v1", metrics={"found": 1, "detail_pages_success": 1, "parsed": 1}, details=[detail(description=None, source_url="")], summary={"valid": 1, "rejected": 0})
assert data_bad["gate_2"]["metrics"]["fields"]["description"]["state"] == "EXPECTED_BUT_NOT_EXTRACTED"
assert data_bad["gate_2"]["metrics"]["fields"]["salary"]["state"] == "UNKNOWN"

unreachable = eight_gates_run_evidence(source="unjobs", adapter_version="fixture:v1", metrics={"found": 2, "detail_pages_success": 0, "parsed": 0}, details=[detail(source_status=0), detail(source_status=0)], summary={"valid": 0, "rejected": 2})
assert unreachable["gate_1"]["reason_code"] == "PROVIDER_UNREACHABLE"
assert unreachable["gate_2"]["status"] == "FAIL"

filtered = eight_gates_run_evidence(source="wwr", adapter_version="fixture:v1", metrics={"found": 1, "detail_pages_success": 1, "parsed": 1}, details=[detail()], summary={"valid": 0, "rejected": 1})
assert filtered["gate_3"]["status"] == "WARNING" and filtered["gate_3"]["reason_code"] == "FILTERED_ROWS"
# Runtime telemetry stores coverage under coverage_state. A bounded listing is
# PARTIAL_COVERAGE (warning), not a provider failure.
bounded = eight_gates_run_evidence(source="unjobs", adapter_version="fixture:v1", metrics={"found": 1, "detail_pages_success": 1, "parsed": 1, "coverage_state": {"complete": False, "stop_reason": "configured_listing_pages"}}, details=[detail()], summary={"valid": 1, "rejected": 0})
assert bounded["gate_1"]["status"] == "WARNING" and bounded["gate_1"]["reason_code"] == "PARTIAL_COVERAGE"
print("PASS eight gates: healthy, data-gap, unknown-not-not-published, provider failure, filters")
